### PR TITLE
Add all sample traces to per sample visualisation beds

### PR DESCRIPTION
**Pull request by jethror1**
- adds all sample traces as anonymised grey lines to per sample visualisation bed files
  - currently when clicked on they will show with a blank label due to an issue in the gCNV track handling in igv.js
  - a separate PR for igv.js has been opened [here](https://github.com/igvteam/igv.js/pull/1917), these will be fully hidden if / when the PR is merged and a new release is made available
- refactored the script since it was a bit messy
- parallelised the output writing as this was the longest part (~ 30 seconds for a CEN run, > 5 minutes for WES), for the same CEN run this now takes <5 seconds
- example with all sample traces added:
image

- - -
This change is [Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20)

## Summary by CodeRabbit

- **New Features**
	- Introduced a new optional parameter `-ikeep_all_sample_traces` in the GATK germline CNV caller app, allowing users to control the visibility of sample traces in output files.
	- Added a `keep_all_sample_traces` parameter in the application configuration to manage sample trace retention.
	- Enhanced the `generate_gcnv_bed.py` script to support conditional inclusion of sample traces based on user preference.
	- Updated the `src/gatk_gcnv_caller.sh` script to include logic for handling sample trace visibility.

- **Bug Fixes**
	- Improved error handling and performance in the BED file generation process through parallel processing.

- **Documentation**
	- Updated README documentation to reflect the new input parameters and their functionalities.

### CODE REVIEW

File: resources/home/dnanexus/generate_gcnv_bed.py [hunk 1]
```diff
@@ -15,103 +19,146 @@ def parse_args():
...
-        )
+            "generate per sample bed file, with just the sample highlighted "
+            "and run level mean and std deviations added"
+        ),
+    )
+    parser.add_argument(
+        "--keep_all_samples",
+        action="store_true",
+        default=True,
+        help=(
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/generate_gcnv_bed.py:43]: good bot, done

File: resources/home/dnanexus/generate_gcnv_bed.py [hunk 2]
```diff
@@ -122,82 +169,181 @@ def generate_per_sample_copy_ratio_dfs(copy_ratio_df):
...
+    ----------
+    bed_file : str
+        bed file to compress and index
+    """
+    subprocess.run(f"bgzip {Path(bed_file)}", shell=True, check=True)
+    subprocess.run(
+        f"tabix -f -S 2 -b 2 -e 3 {Path(bed_file)}.gz",
+        shell=True,
+        check=True,
+    )
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/generate_gcnv_bed.py:308]: I dislike this and there is no risk us running in the app, ignoring this

File: resources/home/dnanexus/generate_gcnv_bed.py [hunk 3]
```diff
@@ -15,103 +19,146 @@ def parse_args():
...
-            "Length of file: {}".format(len(file_df))
-        )
+    for copy_ratio_file in copy_ratio_files:
+        sample, sample_df = read_single_copy_ratio_file(copy_ratio_file)
 
-        # add sample copy ratio to file
-        copy_ratio_df[sample_name] = file_df[sample_name]
+        assert copy_ratio_df[["chr", "start", "end"]].equals(
+            sample_df[["chr", "start", "end"]]
+        ), f"Copy ratio file for {sample} has different intervals "
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/generate_gcnv_bed.py:133]: good bot, have switched to this

File: src/gatk_gcnv_caller.sh [hunk 4]
```diff
@@ -437,9 +437,13 @@ _call_generate_gcnv_bed() {
     denoised_copy_ratio_files=$(find inputs/vcfs/ -name "*_denoised_copy_ratios.tsv")
 
     SECONDS=0
+
+    keep_samples_arg=""
+    if [[ "$keep_all_sample_traces" == "true" ]]; then keep_samples_arg="--keep_all_samples "; fi
+
```
[the remark this answers is not in the document]
Comment (jethror1) [reply src/gatk_gcnv_caller.sh:443]: this is an app input and therefore set when running, ignoring as false positive

**Review by Emma-MacK — CHANGES_REQUESTED** [review]
Have added some suggestions for improving layout, happy that I can follow the logic of the code

> Reviewed 4 of 5 files at r1, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-:-OC82rQC4aY2dg3Up-Uy:b-2vwqdh)* status: all files reviewed, 13 unresolved discussions (waiting on @chrispyatt, @jethror1, and @rg)

___
*[`README.md` line 37 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7rUgW1Ci4tQD358Rm:-OC7rUgW1Ci4tQD358Rn:b-31ww8t) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/README.md#L37)):*
> ```Markdown
> * `-iGermlineCNVCaller_args` (`str`): optional command line arguments for GermlineCNVCaller
> * `-iPostprocessGermlineCNVCalls_args` (`str`): optional command line arguments for PostprocessGermlineCNVCalls
> * `-ikeep_all_sample_traces` (`bool`): controls if to keep all sample traces in per sample visualisation bed files, if true all other samples will be anonymised grey traces
> ```

same pedantic note: should this be "controls whether to keep"? 
___
*[`README.md` line 48 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7sL7a4PN-9q4xOtDz:-OC7sL7a4PN-9q4xOtE-:b-t9wtx8) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/README.md#L48)):*
> ```Markdown
> * `{sample_name}_copy_ratios.gcnv.bed.gz.tbi`: index for the copy ratio bed file
> * `{run_name}_copy_ratios.gcnv.bed.gz`: bed file for the entire run copy ratio visualisation in igv.js
> * `{run_name}_copy_ratios.gcnv.bed.gz`: index for the run copy ratio bed file
> ```

missing ".tbi" and potentially add that this is for the entire run
___
*[`dxapp.json` line 174 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7r-gm65ubFyBDFbej:-OC7r-gm65ubFyBDFbek:b-ku20ix) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/dxapp.json#L174)):*
> ```JSON
>             "optional": true,
>             "default": true,
>             "help": "controls if to keep all samples traces as anonymised grey lines in the per sample visualisation bed file"
> ```

Pedantic: should this be "controls whether to keep"? 
___
*[`resources/home/dnanexus/generate_gcnv_bed.py` line 18 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7tSABEnk2GrJJ9vTE:-OC7tSABEnk2GrJJ9vTF:ba98gl7) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/resources/home/dnanexus/generate_gcnv_bed.py#L18)):*
> ```Python
> from timeit import default_timer as timer
> from typing import Tuple
> ```

Is this blank line needed?
___
*[`resources/home/dnanexus/generate_gcnv_bed.py` line 29 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7th_9CKlx3SXWMLhe:-OC7th_9CKlx3SXWMLhf:blcpcol) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/resources/home/dnanexus/generate_gcnv_bed.py#L29)):*
> ```Python
>         "--copy_ratios",
>         nargs="+",
>         help="all copy ratio files to generate bed file from",
> ```

Is the comma at the end of the line needed? Help is the last section of this argument
___
*[`resources/home/dnanexus/generate_gcnv_bed.py` line 32 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7u5J82rQbKJJoBFmE:-OC7u5J82rQbKJJoBFmF:bfhgnb9) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/resources/home/dnanexus/generate_gcnv_bed.py#L32)):*
> ```Python
>     )
>     parser.add_argument(
>         "--run", help="name of run to prefix output per run bed file with"
> ```

Should the help section be on a new line?
___
*[`resources/home/dnanexus/generate_gcnv_bed.py` line 47 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7uRZY46M1pv4i6zVa:-OC7uRZY46M1pv4i6zVb:b-31ww8t) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/resources/home/dnanexus/generate_gcnv_bed.py#L47)):*
> ```Python
>         action="store_true",
>         help=(
>             "controls if to keep all sample traces in per sample visualisation"
> ```

same pedantic note: should this be "controls whether to keep"?

File: resources/home/dnanexus/generate_gcnv_bed.py [hunk 5]
```diff
@@ -15,103 +19,146 @@ def parse_args():
...
+        intervals and copy ratio values read in from file
+    """
+    with open(copy_ratio_file, encoding="utf-8", mode="r") as fh:
+        # read through file until last line of header with sample name
+        while True:
+            line = fh.readline()
+            if line.startswith("@RG"):
+                sample_name = line.split("SM:")[1].strip()
+                break
+
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/generate_gcnv_bed.py]: Done.

File: resources/home/dnanexus/generate_gcnv_bed.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/generate_gcnv_bed.py]: Done.

File: resources/home/dnanexus/generate_gcnv_bed.py [hunk 6]
```diff
@@ -15,103 +24,172 @@ def parse_args():
...
+            copy_ratio_file,
+            sep="\t",
+            comment="@",
+            header=0,
+            names=["chr", "start", "end", sample_name],
         )
+    except pd.errors.EmptyDataError:
+        raise ValueError(f"No data found in file: {copy_ratio_file}")
+    except Exception as e:
+        raise ValueError(f"Error reading file {copy_ratio_file}: {str(e)}")
```
[the remark this answers is not in the document]
Comment (jethror1) [reply resources/home/dnanexus/generate_gcnv_bed.py:108]: Done.

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20)* status: all files reviewed, 10 unresolved discussions (waiting on @chrispyatt, @coderabbitai[bot], @Emma-MacK, and @rg)

___
*[`dxapp.json` line 174 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7r-gm65ubFyBDFbej:-OC84rXv9t_Ls3Lq6ROv:bb6un1j) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/dxapp.json#L174)):*
<details><summary><i>Previously, Emma-MacK wrote…</i></summary><blockquote>

Pedantic: should this be "controls whether to keep"?
</blockquote></details>

Done
___
*[`README.md` line 48 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7sL7a4PN-9q4xOtDz:-OC84qUPBFZNWX2TCCql:b-896fix) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/README.md#L48)):*
<details><summary><i>Previously, Emma-MacK wrote…</i></summary><blockquote>

missing ".tbi" and potentially add that this is for the entire run
</blockquote></details>

Done.
___
*[`resources/home/dnanexus/generate_gcnv_bed.py` line 18 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7tSABEnk2GrJJ9vTE:-OC84zsOCWgDWmKVZ5cj:b3kgry7) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/resources/home/dnanexus/generate_gcnv_bed.py#L18)):*
<details><summary><i>Previously, Emma-MacK wrote…</i></summary><blockquote>

Is this blank line needed?
</blockquote></details>

si, 3rd party imports should be separated from std lib (for pep8ness)
___
*[`resources/home/dnanexus/generate_gcnv_bed.py` line 29 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7th_9CKlx3SXWMLhe:-OC85foGBob7TFCNQTxU:b-3eicpz) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/resources/home/dnanexus/generate_gcnv_bed.py#L29)):*
<details><summary><i>Previously, Emma-MacK wrote…</i></summary><blockquote>

Is the comma at the end of the line needed? Help is the last section of this argument
</blockquote></details>

this is a feature of the Black formatter that I have enabled, see:
- https://test-black.readthedocs.io/en/style-guide/style_guide/trailing_commas.html
- https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#labels-magic-trailing-comma
___
*[`resources/home/dnanexus/generate_gcnv_bed.py` line 32 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7u5J82rQbKJJoBFmE:-OC85mxR6HbNOw3UP7qI:b-nhqy4o) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/resources/home/dnanexus/generate_gcnv_bed.py#L32)):*
<details><summary><i>Previously, Emma-MacK wrote…</i></summary><blockquote>

Should the help section be on a new line?
</blockquote></details>

again this is Black formatting, it will stick to one line where it is under the line length
___
*[`resources/home/dnanexus/generate_gcnv_bed.py` line 47 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7uRZY46M1pv4i6zVa:-OC863NdBw2OAlVqA032:bb6un1j) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/resources/home/dnanexus/generate_gcnv_bed.py#L47)):*
<details><summary><i>Previously, Emma-MacK wrote…</i></summary><blockquote>

same pedantic note: should this be "controls whether to keep"?
</blockquote></details>

Done

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20)* status: 2 of 5 files reviewed, 10 unresolved discussions (waiting on @chrispyatt, @coderabbitai[bot], @Emma-MacK, and @rg)

___
*[`resources/home/dnanexus/generate_gcnv_bed.py` line 29 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7th_9CKlx3SXWMLhe:-OC86Pll9zW2Z4hq5ln8:b-nv4taj) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/resources/home/dnanexus/generate_gcnv_bed.py#L29)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

this is a feature of the Black formatter that I have enabled, see:
- https://test-black.readthedocs.io/en/style-guide/style_guide/trailing_commas.html
- https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#labels-magic-trailing-comma
</blockquote></details>

also this in case you've not seen Black before: https://binfx.slack.com/archives/G01HV61U4BX/p1726827355353159

**Review by Emma-MacK — APPROVED** [review]
> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20)* status: all files reviewed, 6 unresolved discussions (waiting on @chrispyatt, @jethror1, and @rg)

___
*[`resources/home/dnanexus/generate_gcnv_bed.py` line 18 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7tSABEnk2GrJJ9vTE:-OC87pnkCO970iwl6KG2:b-79wats) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/resources/home/dnanexus/generate_gcnv_bed.py#L18)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

si, 3rd party imports should be separated from std lib (for pep8ness)
</blockquote></details>

Ah interesting, thank you!
___
*[`resources/home/dnanexus/generate_gcnv_bed.py` line 29 at r4](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20#-OC7th_9CKlx3SXWMLhe:-OC880d30L5mu9k8d4aB:b-9lpq8k) ([raw file](https://github.com/eastgenomics/eggd_GATKgCNV_call/blob/185d896ff0e0ef3b117c080e85210b47990db08d/resources/home/dnanexus/generate_gcnv_bed.py#L29)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

also this in case you've not seen Black before: https://binfx.slack.com/archives/G01HV61U4BX/p1726827355353159
</blockquote></details>

Amazing, thanks!

**Review by chrispyatt — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_GATKgCNV_call/20)* status: all files reviewed, 6 unresolved discussions (waiting on @coderabbitai[bot], @jethror1, and @rg)

**Pull request merged**